Renderer file logging transport

samitiilikainen opened the pull request:
Add file logging to renderer, writing separate log files for renderer main frame and each cluster frame.

Closes file handles (otherwise left open by Winston) on main and cluster frame refresh and when cluster frame is closed through cluster disconnect.

Related to lensapp/support-lens-extension#118

samitiilikainen committed:
Renderer file logging transport

Add file logging to renderer, writing separate log files for renderer main frame and each cluster frame.

Related to lensapp/support-lens-extension#118

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

samitiilikainen committed:
Switch renderer file log level to info

There is too much noise on debug level from api responses etc

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

samitiilikainen committed:
Unmount on pagehide instead of beforeunload

It seems cluster onbeforeunload is not triggered when iframe is removed from dom by the parent (when disconnecting a cluster).

While on root/main frame the beforeunload event does work, it seems to be adviced to use pagehide instead.

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

samitiilikainen committed:
Close log files on unmount

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

samitiilikainen committed:
Improve file handle closing in different situations

This should cover reloading main and cluster frames and closing cluster frame throught disconnecting cluster. No file handles should be left open now.

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

jim-docker commented on packages/core/src/renderer/logger/file-transport.injectable.ts:
```diff
@@ -0,0 +1,44 @@
+import currentlyInClusterFrameInjectable from "../routes/currently-in-cluster-frame.injectable";
+import { getClusterIdFromHost } from "../utils";
+
+const rendererFileLoggerTransportInjectable = getInjectable({
```
usually the filename matches the injectable name/id ?

jim-docker commented on packages/core/src/renderer/frames/cluster-frame/init-cluster-frame/init-cluster-frame.ts:
```diff
@@ -73,11 +76,14 @@ export const initClusterFrame = ({
+    });
+
+    window.addEventListener("beforeunload", onCloseFrame, { capture: true });
+    window.addEventListener("pagehide", onCloseFrame, { capture: true });
```
Does this mean logging to file stops if the cluster frame is not in view?  If I disconnect a cluster while on the catalog cluster list, any disconnect logging is missed?

samitiilikainen replied:
Not all logging. It stops cluster frame logging right before the iframe is removed from DOM (when also all code execution in that iframe should stop). Stopping logging of the cluster frame does not affect logging done by the renderer main frame (catalog page).

jim-docker commented on packages/core/src/renderer/logger/file-transport.injectable.ts:
```diff
@@ -0,0 +1,44 @@
+      dirname: di.inject(directoryForLogsInjectable),
+      maxsize: 1024 * 1024,
+      maxFiles: 0,
+      tailable: true,
```
Not sure what the expectation is here, but I tested with a file that was just under 1M in size and observed that when the file reached 1M nothing more was written to the file or any other file, and nothing more was written to the developer tools console either??

https://user-images.githubusercontent.com/40840436/215165042-11ae55cf-3529-4fb0-91a4-9baa9be15dea.mov

This is on Ubuntu 22.04. I artificially padded the file with logs to get it near 1M in size.

samitiilikainen replied:
Thanks! Expectation was that it would keep writing logs and older lines would be lost but I was mistaken. I'll increase the maxFiles to allow it to properly rotate files.

jim-docker reviewed (commented):
I noticed we have many `console.warn` etc calls still in the core code. They show up in the dev tools console but not in the log file of course. Out of scope for this PR, just noting it.

samitiilikainen committed:
Fix renderer log rotation

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

samitiilikainen committed:
Switch back to beforeunload in root frame

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

jim-docker reviewed (dismissed):
Thanks for doing this!

(users with many many clusters may notice large disk usage by Lens due to all the new log files)

samitiilikainen committed:
Remove capturing phase event usage

Does not seem to be needed for log files to be successfully closed and caused integration tests to fail.

Signed-off-by: Sami Tiilikainen <97873007+samitiilikainen@users.noreply.github.com>

Nokel81 reviewed (approved):
Finally looks like this is green!
